### PR TITLE
Support admin requests in the fake RP

### PR DESCRIPTION
**Pull request by kargakis**
This PR creates a new endpoint in the fake RP where the admin API is served
and enables handling of admin requests.

@jim-minter @charlesakalugwu 

Part of AZURE-119

### CODE REVIEW

File: cmd/createorupdate/createorupdate.go [hunk 1]
```diff
@@ -289,6 +293,9 @@ func main() {
 		s := fakerp.NewServer(log, conf.ResourceGroup, fakeRpAddr, conf)
 		go s.ListenAndServe()
 	}
+	if *isAdmin {
+		rpURL += "/admin"
```
Comment (kargakis) [inline cmd/createorupdate/createorupdate.go]: Instead of doing this we should probably create a client in pkg/admin/api and use that here instead of using `v20180930preview.NewOpenShiftManagedClustersClientWithBaseURI` below.

File: pkg/fakerp/server.go [hunk 2]
```diff
@@ -52,6 +53,7 @@ func NewServer(log *logrus.Entry, resourceGroup, address string, c *Config) *Ser
 func (s *Server) ListenAndServe() {
 	// TODO: match the request path the real RP would use
 	http.Handle("/", s)
+	http.Handle("/admin", s)
```
Comment (kargakis) [inline pkg/fakerp/server.go]: Unclear how this endpoint will look like in prod.

**Comment by kargakis** [comment]
@jim-minter @charlesakalugwu @mjudeikis ptal, I have updated the fake RP to support accepting admin requests. Follow-up PRs to add admin validation and enable createorupdate to send admin requests.

**Comment by kargakis** [comment]
> I personally dont like how request path is being implemented here, (`/` vs `/admin`)

@jim-minter and I agreed on a separate endpoint for the admin api, I am open to better suggestions on the paths :)

File: pkg/fakerp/server.go [hunk 3]
```diff
@@ -249,70 +267,121 @@ func (s *Server) handlePut(w http.ResponseWriter, req *http.Request) {
...
+		http.Error(w, resp, http.StatusBadRequest)
+		return
 	}
 
-	if _, err := CreateOrUpdate(ctx, oc, s.log, config); err != nil {
-		s.writeState(v20180930preview.Failed)
-		resp := "400 Bad Request: Failed to apply request"
-		s.log.Debugf("%s: %v", resp, err)
+	// apply the request
+	cs, err = createOrUpdate(ctx, s.log, cs, oldCs, config)
```
[the remark this answers is not in the document]
Comment (mjudeikis) [reply pkg/fakerp/server.go]: What is the long run plan? Is the plan to "abstract" all `Update` queries behind `createOrUpdate` function, or we will split those to smaller ones if needed? In example `MigrateDisk` ?

**Comment by mjudeikis** [comment]
> > I personally dont like how request path is being implemented here, (`/` vs `/admin`)
> 
> @jim-minter and I agreed on a separate endpoint for the admin api, I am open to better suggestions on the paths :)

separate endpoint is must, but each time I see something like:
```
isAdminRequest := strings.HasPrefix(req.URL.Path, "/admin")
```
I think about "path router" for http. But as I say, 2 Paths - fine. when we start doing this for 4 - we might need to think about router

**Comment by kargakis** [comment]
Oh yeah, that can be a follow-up enhancement.

**Comment by kargakis** [comment]
connection timeout

/retest

File: pkg/fakerp/server.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (kargakis) [reply pkg/fakerp/server.go]: I think we are going to have separate functions to describe common flows like key rotation or backup+restore.

File: pkg/fakerp/config.go [hunk 4]
```diff
@@ -76,7 +76,7 @@ func NewConfig(log *logrus.Entry, needRegion bool) (*Config, error) {
 	return &c, nil
 }
 
-func GetPluginConfig() (*api.PluginConfig, error) {
+func GetPluginConfig(isAdmin bool) (*api.PluginConfig, error) {
 	tc := api.TestConfig{
 		RunningUnderTest:      os.Getenv("RUNNING_UNDER_TEST") != "",
```
Comment (charlesakalugwu) [inline pkg/fakerp/config.go]: do you need to rebase? this line is `os.Getenv("RUNNING_UNDER_TEST") == "true"` in master

Comment (kargakis) [reply pkg/fakerp/config.go]: Not necessary, my branch is not on top of HEAD but there is no conflict.

**Comment by kargakis** [comment]
https://github.com/openshift/openshift-azure/issues/814

/retest

**Comment by kargakis** [comment]
https://github.com/openshift/openshift-azure/issues/863

/retest

File: pkg/api/plugin.go [hunk 5]
```diff
@@ -73,8 +73,8 @@ type PluginConfig struct {
 	SyncImage       string
 	AcceptLanguages []string
 	GenevaConfig    GenevaConfig
-
-	TestConfig TestConfig
+	IsAdmin         bool
```
Comment (jim-minter) [inline pkg/api/plugin.go]: why is this here?  surely it shouldn't be?

File: test/clients/azure/osa.go [hunk 6]
```diff
@@ -44,13 +43,12 @@ func (cli *Client) UpdateOSACluster(ctx context.Context, external *v20180930prev
...
-	logger := logrus.NewEntry(logrus.StandardLogger())
 
 	var oc *v20180930preview.OpenShiftManagedCluster
 	var err error
 	// simulate the API call to the RP
-	if oc, err = fakerp.CreateOrUpdate(ctx, external, logger, config); err != nil {
-		return nil, err
-	}
-	return oc, nil
+	//if oc, err = fakerp.CreateOrUpdate(ctx, external, logger, config); err != nil {
```
Comment (jim-minter) [inline test/clients/azure/osa.go]: why is this commented out?  is a middle ground necessary before this entire function gets deleted?

Comment (kargakis) [reply test/clients/azure/osa.go]: Waiting on https://github.com/openshift/openshift-azure/pull/880 to merge, then I am going to fix this.

File: pkg/api/plugin.go (hunk 5 above)
Comment (kargakis) [reply pkg/api/plugin.go]: Hrm, shouldn't be here - I have split validation into its own PR, see https://github.com/openshift/openshift-azure/pull/854

File: pkg/fakerp/server.go [hunk 7]
```diff
@@ -19,9 +18,12 @@ import (
 	"github.com/ghodss/yaml"
 	"github.com/sirupsen/logrus"
 
+	"github.com/openshift/openshift-azure/pkg/api"
 	internalapi "github.com/openshift/openshift-azure/pkg/api"
```
Comment (charlesakalugwu) [inline pkg/fakerp/server.go]: remove one of the two imports

File: pkg/fakerp/server.go [hunk 8]
```diff
@@ -126,44 +118,61 @@ func (s *Server) validate(w http.ResponseWriter, r *http.Request) bool {
...
+// be consistent with how the RP runs in production so we need
+// to restore the internal state of the cluster from the
+// filesystem.
+func (s *Server) restore(path string) error {
+	dataDir, err := FindDirectory(path)
+	if err != nil {
+		return err
 	}
-	data, err := ioutil.ReadFile(path)
+	data, err := ioutil.ReadFile(dataDir)
```
Comment (charlesakalugwu) [inline pkg/fakerp/server.go]: you should read the containerservice.yaml and not path (which is most likely _data)

File: pkg/api/validate.go [hunk 9]
```diff
@@ -501,3 +518,10 @@ func (v *Validator) validateProvisioningState(ps ProvisioningState) (errs []erro
 	}
 	return
 }
+
+func (v *Validator) validateUpdateConfig(internalConfig, adminConfig Config) (errs []error) {
+	if !reflect.DeepEqual(internalConfig, adminConfig) {
```
Comment (charlesakalugwu) [inline pkg/api/validate.go]: wouldn't this be a problem for key rotation which needs to have certain fields in the Config invalidated and regenerated before using that in the update?

Comment (kargakis) [reply pkg/api/validate.go]: Yes, we agreed that in the first cut of this, we should invalidate every change. Your changes should enable validating certificate updates.

File: pkg/fakerp/server.go (hunk 8 above)
Comment (kargakis) [reply pkg/fakerp/server.go]: good catch, fixed

File: pkg/fakerp/server.go (hunk 7 above)
Comment (kargakis) [reply pkg/fakerp/server.go]: fixed
